hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/fsdataset/impl/ReplicaMap.java

```diff
@@ -166,25 +167,24 @@ void addAll(ReplicaMap other) {
   /**
    * Merge all entries from the given replica map into the local replica map.
    */
-  void mergeAll(ReplicaMap other) {
+  void mergeAll(ReplicaMap other) throws IOException {
     Set<String> bplist = other.map.keySet();
     for (String bp : bplist) {
       checkBlockPool(bp);
       try (AutoCloseDataSetLock l = lockManager.writeLock(LockLevel.BLOCK_POOl, bp)) {
         LightWeightResizableGSet<Block, ReplicaInfo> replicaInfos = other.map.get(bp);
         LightWeightResizableGSet<Block, ReplicaInfo> curSet = map.get(bp);
+        if (curSet == null) {
+          // Can't find the block pool id in the replicaMap. Maybe it has been removed.
```

Comment:
just for myself learning~~ is it possible we can't find block pool id from the map, but the block pool is not removed yet? (say, if something went wrong during the deleting and the pool was not removed successfully?)

Reply:
@DaveTeng0 Thanks for your review.

> is it possible we can't find block pool id from the map, but the block pool is not removed yet?

I think it's impossible. If you find some cases, please share.

1. The `mergeAll` method only been used by the `activateVolume` method with a `synchronized` lock.
2. The `shutdownBlockPool` method only been used by the `shutdownBlockPool` method with a `synchronized` lock, which beed added in this PR.
3. The `addBlockPool` method always initialize an empty `LightWeightResizableGSet` in the global `ReplicaMap` for this block pool.

Reply:
Thanks Zander for your detail explanation! got this!!